c10/cuda/CUDAAssert.cpp

```diff
@@ -0,0 +1,248 @@
+#include <c10/cuda/CUDAGuard.h>
+
+#include <stdio.h>
+#include <cstring>
+#include <map>
+#include <memory>
+#include <sstream>
+#include <stdexcept>
+#include <unordered_map>
+
+#include "CUDAAssert.h"
+
+namespace c10 {
+namespace cuda {
+
+struct FileErrorRange {
+  std::string file;
+  int32_t first;
+  int32_t last;
+};
+
+#define FILE_ERROR_RANGE(file, tag)                 \
+  {                                                 \
+    (int32_t) tag::FIRST, {                         \
+      file, (int32_t)tag::FIRST, (int32_t)tag::LAST \
+    }                                               \
+  }
+
+static std::map<int32_t, FileErrorRange> error_range_file_table{
+    FILE_ERROR_RANGE("MultinomialKernel.cu", AssertTag::MultinomialKernel),
+    FILE_ERROR_RANGE("BinaryOpsKernel.cu", AssertTag::BinaryOpsKernel),
+    FILE_ERROR_RANGE("IndexKernel.cu", AssertTag::IndexKernel),
+    FILE_ERROR_RANGE("THCTensorIndex.cu", AssertTag::THCTensorIndex),
+    FILE_ERROR_RANGE("ClassNLLCriterion.cu", AssertTag::ClassNLLCriterion),
+    FILE_ERROR_RANGE("Distributions.cu", AssertTag::Distributions),
+    FILE_ERROR_RANGE("EmbeddingBag.cu", AssertTag::EmbeddingBag),
+    FILE_ERROR_RANGE("FractionalMaxPoo2d.cu", AssertTag::FractionalMaxPoo2d),
+    FILE_ERROR_RANGE(
+        "SpatialClassNLLCriterion.cu",
+        AssertTag::SpatialClassNLLCriterion)};
+
+// global assert error output format table
+static std::unordered_map<int32_t, const char*> assert_format_table{
+    // MultinomialKernel.cu:
+    {(int32_t)AssertTag::MultinomialKernel::_004,
+     "invalid multinomial distribution (encountering probability entry < 0)"},
+    {(int32_t)AssertTag::MultinomialKernel::_005,
+     "invalid multinomial distribution (encountering probability entry = infinity)"},
+    {(int32_t)AssertTag::MultinomialKernel::_006,
+     "invalid multinomial distribution (encountering probability entry = NaN)"}};
+
+template <typename T>
+static std::string formatToken(const std::string& fmt, const T& data) {
+  int ret = snprintf(nullptr, 0, fmt.c_str(), data);
+  if (ret < 0) {
+    throw std::runtime_error("Calculating token output size failed.");
+  }
+
+  if (ret == 0) {
+    return std::string{};
+  }
+
+  size_t size = ret + 1; // ret does not include the terminating '\0'
+  auto buff_ptr = std::unique_ptr<char[]>(new char[size]);
+  ret = snprintf(buff_ptr.get(), size, fmt.c_str(), data);
+  if (ret < 0 || ret >= size) {
+    throw std::runtime_error("Formatting token failed.");
+  }
+
+  return std::string(buff_ptr.get());
+}
+
+static std::string formatAssertOutput(const char* format, char* data) {
```

Comment:
IIRC we don't need this anymore either, right?

Reply:
It is used for formatting the output strings, potentially every entry in the format string table could use the full fprintf set for arguments captured at the assert-call-site, here https://github.com/pytorch/pytorch/blob/54bfaf896f37d49bdd05517c460edfa8fc3dcb4a/c10/cuda/CUDAAssert.cpp#L43-L50

Currently the only place for format-strings is:
https://github.com/pytorch/pytorch/blob/54bfaf896f37d49bdd05517c460edfa8fc3dcb4a/c10/cuda/CUDAAssert.cpp#L189

of which you can see the argument capturing here:
https://github.com/pytorch/pytorch/blob/54bfaf896f37d49bdd05517c460edfa8fc3dcb4a/aten/src/ATen/native/cuda/Assert.cuh#L201-L203
(every C10_ASSERT can theoretically use this argument capturing feature)

Reply:
Sounds good. Docs? :)